teos/src/responder.rs

```diff
@@ -539,67 +393,47 @@ impl chain::Listen for Responder {
-                .iter()
-                .map(|uuid| (*uuid, self.trackers.lock().unwrap()[uuid].user_id))
-                .collect();
-            self.delete_trackers(
-                &rejected_trackers,
-                &self
-                    .gatekeeper
-                    .delete_appointments_from_memory(&trackers_to_delete_gk),
-                DeletionReason::Rejected,
-            );
+        // Delete trackers completed at this height
+        let completed_trackers =
+            self.check_confirmations(txdata.iter().map(|(_, tx)| tx.txid()).collect(), height);
+        self.gatekeeper
+            .delete_appointments(completed_trackers, true);
 
-            // Remove all receipts created in this block
-            self.carrier.lock().unwrap().clear_receipts();
+        // Rebroadcast those transactions that need to
+        let rejected_trackers = self.rebroadcast_stale_txs(height);
+        self.gatekeeper
+            .delete_appointments(rejected_trackers, false);
 
-            if self.trackers.lock().unwrap().is_empty() {
-                log::info!("No more pending trackers");
-            }
-        }
+        // Remove all receipts created in this block
+        self.carrier.lock().unwrap().clear_receipts();
     }
 
     /// Handles reorgs in the [Responder].
     fn block_disconnected(&self, header: &BlockHeader, height: u32) {
         log::warn!("Block disconnected: {}", header.block_hash());
-        self.carrier.lock().unwrap().update_height(height);
+
+        let dbm = self.dbm.lock().unwrap();
+        let mut carrier = self.carrier.lock().unwrap();
+
+        carrier.update_height(height);
         self.tx_index
             .lock()
             .unwrap()
             .remove_disconnected_block(&header.block_hash());
 
-        for tracker in self.trackers.lock().unwrap().values_mut() {
+        // TODO: Edit this db call to return only `confirmed` trackers in `height`.
+        for (uuid, tracker_summary) in dbm.load_tracker_summaries() {
             // The transaction has been unconfirmed. Flag it as reorged out so we can rebroadcast it.
-            if tracker.status == ConfirmationStatus::ConfirmedIn(height) {
-                tracker.status = ConfirmationStatus::ReorgedOut;
+            if tracker_summary.status == ConfirmationStatus::ConfirmedIn(height) {
+                let tracker = dbm.load_tracker(uuid).unwrap();
+                // Rebroadcast the dispute and the penalty transactions.
+                // We don't care about their status for now. If they won't be accepted, they will
+                // be cleaned up later when they miss too many confirmations.
+                carrier.send_transaction(&tracker.dispute_tx);
+                carrier.send_transaction(&tracker.penalty_tx);
+                // Update the tracker status.
+                dbm.update_tracker_status(uuid, &ConfirmationStatus::InMempoolSince(height))
+                    .unwrap();
```

Comment:
The on connection way make more sense actually (but don't think this is so bad tho).
This is trying to avoid having some struct in memory to hold these reorged blocks until we fully disconnect the weak chain and start connecting the strong one.
Would be complex also to add a representation for the reorg case if we were to use the DB for storing this state.

What this currently does is:
- When a block is disconnected, we will publish the dispute and the penalty right away (not caring if they were rejected or not, since they should actually be published on our way up (connections) and not down (disconnections))
- After some time, they they are still not yet accepted into the chain in `CONFIRMATIONS_BEFORE_RETRY`, we will try again to push them. This time, removing them if they are invalid.

Reply:
This doesn't change the old logic functionality-wise tho.
If some x blocks where disconnected, the very next connected block should republish all the reorged txs we care about.
This does the same but republishing everything on the way down.
The thing is, we publish them anyway without knowing if they are actually on-chain/in mempool since we don't have access to these info in the disconnection time.